cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/BasicWorkflowToken.java

```diff
@@ -87,15 +101,16 @@ void mergeToken(BasicWorkflowToken other) {
 
         for (NodeValue otherNodeValue : otherTokenValueMapForScopeEntry.getValue()) {
           boolean otherNodeValueExist = false;
-          for (NodeValue thisNodeValue :
-            thisTokenValueMapForScope.get(otherTokenValueMapForScopeEntry.getKey())) {
+          for (NodeValue thisNodeValue : thisTokenValueMapForScope.get(otherKey)) {
             if (thisNodeValue.equals(otherNodeValue)) {
               otherNodeValueExist = true;
               break;
             }
           }
           if (!otherNodeValueExist) {
-            thisTokenValueMapForScope.get(otherTokenValueMapForScopeEntry.getKey()).add(otherNodeValue);
+            ensureEnoughMemoryAvailable(otherKey, otherNodeValue.getValue());
+            thisTokenValueMapForScope.get(otherKey).add(otherNodeValue);
+            updateSize(otherKey, otherNodeValue.getValue());
```

Comment:
Why not encapsulate this three into a method like `addNodeValue(Map<String, List<NodeValue>> map, String key, NodeValue nodeValue)` that would do check and update size and adding to the list?
```